modules/dataplex-datascan/variables.tf

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "data" {
+  description = "The data source for DataScan. The source can be either a Dataplex `entity` or a BigQuery `resource`."
+  type = object({
+    entity   = optional(string)
+    resource = optional(string)
+  })
+  validation {
+    condition     = length([for k, v in var.data : v if contains(["resource", "entity"], k) && v != null]) == 1
+    error_message = "Datascan data must specify one of 'entity', 'resource'."
+  }
+}
+
+variable "data_profile_spec" {
+  description = "DataProfileScan related setting. Variables description are provided in https://cloud.google.com/dataplex/docs/reference/rest/v1/DataProfileSpec."
+  default     = null
+  type = object({
+    sampling_percent = optional(number)
+    row_filter       = optional(string)
+  })
+}
+
+variable "data_quality_spec" {
+  description = "DataQualityScan related setting. Variables description are provided in https://cloud.google.com/dataplex/docs/reference/rest/v1/DataQualitySpec."
+  default     = null
+  type = object({
+    sampling_percent = optional(number)
+    row_filter       = optional(string)
+    rules = list(object({
```

Comment:
are expectations mutually exclusive or can they be specified together in the same rule?

Reply:
each rule object may only have one expectation. I've created a precondition check for this https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/pull/1545/files#diff-c2745954b00105f8387b607ccbe9413b234e75a8c7b20dad05c713e91e0cc86cR194

Reply:
ah right, sorry I was expecting to see it as a validation rule in the variable... it's ok, we can move it to validation later

Reply:
It was in the variable before :) Now with the additional option to provide the spec as a file with `var.data_quality_spec_file`, I moved the validation step into the module so it is applied regardless of whether the user supplied the input via `var.data_quality_spec` or `var.data_quality_spec_file`

Reply:
okok ignore me :)